File: 98-validate-binary-search-tree/validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool valid(TreeNode* root,TreeNode* left,TreeNode* right){
        if(root==NULL) return true;

        if(left!=NULL && root->val<=left->val || (right!=NULL && root->val>=right->val)) return false;
        return valid(root->left,left,root) && valid(root->right,root,right);

    }
    bool isValidBST(TreeNode* root) {
        if(root==NULL) return true;

        return valid(root,NULL,NULL);
    }
};

File: 98-validate-binary-search-tree/validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean dfs(TreeNode root,Integer left,Integer right){
        if(root==null) return true;

        if((left!=null && left>=root.val) || (right!=null && right<=root.val)) return false;
        return dfs(root.left,left,root.val) && dfs(root.right,root.val,right);
    }
    public boolean isValidBST(TreeNode root) {
        if(root==null) return true;
        return dfs(root,null,null);
    }
}